### PR TITLE
3.next - Add loose emptiness handling for `HasMany` associations.

### PR DESCRIPTION
**Pull request by ndm2**
Basically copied the functionality and tests from `BelongsToMany` with slight modifications and additions. Also refactored the `saveAssociated` method a little bit, making it easier to test and to read.

I wasn't sure about the rules regarding backwards compatibility for protected methods, if breaking changes were allowed, I would have changed `_unlinkAssociated()` to drop the `$entity` (unused) and `$target` arguments.

Refs #9474

### CODE REVIEW

File: tests/TestCase/ORM/Association/HasManyTest.php [hunk 1]
```diff
@@ -740,4 +740,216 @@ public function testLinkUsesSingleTransaction()
...
+            ->setConstructorArgs(['comments', ['sourceTable' => $table]])
+            ->getMock();
+        $assoc->setSaveStrategy(HasMany::SAVE_APPEND);
+
+        $assoc
+            ->expects($this->never())
+            ->method('_unlinkAssociated');
+        $assoc
+            ->expects($this->never())
+            ->method('_saveTarget');
```
Comment (markstory) [inline tests/TestCase/ORM/Association/HasManyTest.php]: Instead of mocks integrating with the db and checking the db state afterwards has been resulting in easier to read and easier to maintain tests lately.

Comment (ndm2) [inline tests/TestCase/ORM/Association/HasManyTest.php]: @markstory You mean in this horrendously large beast called the "table test", also know as the unholy slayer of sanity? :wink:

Jokes aside, I could surely switch those to integration tests if that is preferred / considered sufficient. Without mocking, I imagine it could be hard to differentiate between bailing out early on empty sets, and having nothing to unlink/persist.

Comment (lorenzo) [inline tests/TestCase/ORM/Association/HasManyTest.php]: You could maybe do that by asserting that the set you are passing is not empty

Comment (ndm2) [inline tests/TestCase/ORM/Association/HasManyTest.php]: Maybe I misunderstood what you're suggesting, but I actually need to test this functionality with empty sets :) I probably overcomplicate things in my head though, I'll give it a try tomorrow and see if it's actually an issue at all.

Comment (markstory) [inline tests/TestCase/ORM/Association/HasManyTest.php]: @ndm Users of the ORM care about the persistence results not which protected methods are called though. I think the 'integration' tests could continue to exist in this file. I recently change a bunch of association class tests to not use mocks and it greatly simplified the test code.

Comment (ndm2) [inline tests/TestCase/ORM/Association/HasManyTest.php]: @markstory Of course, my concern was just that in a sitation where the method can return one and the same value for different internal actions, it wouldn't be possible to determine that the result, even though being the expected one (an empty array, for an empty array as input), is the result of internals working as expected or not.

But that's kind of a personal habit, I tend to over-test things, and giving it a try it seems that my concern was unsubstantiated, looks like it's enough to test the possibly existing associated records, and that the input property value equals the output property value :)

I've updated the tests.

**Review by markstory — COMMENTED** [review]
Looks good to me.
